### PR TITLE
QTI-68: Added custom content type mappings for the non-standard TIMS …

### PR DESCRIPTION
**Pull request by bdrainer**
…files extensions.  The mappings are configurable using the application yamls.

Tika looks cool but I need more time figuring out how to customize it correctly.

### CODE REVIEW

File: application.yml [hunk 1]
```diff
@@ -15,6 +15,12 @@
 server:
   port: 8081
 
+tims-content-types:
+  mappings:
+    "brf": "application/braille"
+    "prn": "application/braille"
+    "qrx": "application/xml"
```
Comment (bdrainer) [inline application.yml]: This is a first pass at custom file extension to content type mappings.  Putting them in a config file means adding/updating these only requires a bounce of the app and not a build/release of the app.

File: src/main/java/org/opentestsystem/ap/ims/service/impl/ItemEntityServiceImpl.java [hunk 2]
```diff
@@ -73,6 +78,15 @@ public void initialize() {
                 log.error(format("Error creating folder %s.  Message: %s", path, e.getMessage()), e);
             }
         }
+
+        this.customMediaTypes = new HashMap<>();
+
+        this.customMediaTypes = this.contentTypeMappings.getMappings().entrySet().stream()
+            .collect(
+                Collectors.toMap(e ->
+                    Pattern.compile(".*[.]" + e.getKey()), e -> org.apache.tika.mime.MediaType.parse(e.getValue())
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ims/service/impl/ItemEntityServiceImpl.java]: Here is the conversion from the properties in the yaml to the actual Tika map.  Tika uses a Pattern as the map key and the content type value as the map value.

File: application.yml (hunk 1 above)
Comment (wtritch) [reply application.yml]: I like it =)

File: src/main/java/org/opentestsystem/ap/ims/service/impl/ItemEntityServiceImpl.java [hunk 3]
```diff
@@ -162,4 +164,46 @@ private ItemResource buildItemResource(S3Object s3Object, String fileName) {
...
+        return tempFilePath;
+    }
+
+    private String determineMediaType(Path filePath, String fileName) throws IOException {
+        Metadata metadata = new Metadata();
+        metadata.set(Metadata.RESOURCE_NAME_KEY, fileName);
+
+        // first look for our custom ones
+        String mediaType = this.determineMediaType(
+            new Tika(new NameDetector(this.customMediaTypes)),
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/ims/service/impl/ItemEntityServiceImpl.java]: Hm... does instantiating a Tika consume resources?  Is it something thread-safe that we only need one of?

Comment (tims-fairway-system) [reply src/main/java/org/opentestsystem/ap/ims/service/impl/ItemEntityServiceImpl.java]: I made it a member of the class.  I think it is thread-safe.

**Pull request merged**